knit/src/Knit/Autocompletion.hs

```diff
@@ -0,0 +1,251 @@
+module Knit.Autocompletion
+       ( suggestions
+       ) where
+
+import Control.Monad.Writer.Strict (runWriter)
+import Data.List (isPrefixOf)
+import Data.Loc (loc, spanFromTo)
+import Data.Maybe (mapMaybe)
+import Data.Semigroup ((<>))
+import qualified Data.Text as T
+import Data.Text.Buildable (build)
+import Data.Text.Lazy (unpack)
+import Data.Text.Lazy.Builder (toLazyText)
+import Text.Earley (fullParses)
+
+import Knit.Argument
+import Knit.FormattedExprExt
+import Knit.Name
+import Knit.Parser
+import Knit.ParseTreeExt
+import Knit.Prelude
+import Knit.Printer
+import Knit.Procedure
+import Knit.Syntax
+import Knit.Tokenizer
+
+suggestions
+  :: forall components proxy.
+     ( KnownSpine components
+     , AllConstrained (ComponentTokenizer components) components
+     , AllConstrained (ComponentTokenToLit components) components
+     , AllConstrained (ComponentCommandProcs components) components
+     , AllConstrained ComponentPrinter components
+     )
+  => proxy components
+  -> T.Text
+  -> [T.Text]
+suggestions _ cmd =
+  let
+    cmd' = "(" <> cmd
```

Comment:
I'd like to see this parentheses business gone. Also, I don't like computations that involve spans.

Instead, let's change the tokenizer to return

```
newtype Skipped = Skipped String
(Skipped, [(SSpan, (Token components, Skipped))])
```

which is easy to achieve if we redefine `pSkip` as

```
pSkip = Skipped <$> many spaceChar
```